bgpd/bgp_labelpool.c

```diff
@@ -1555,3 +1678,77 @@ void bgp_lp_vty_init(void)
 	install_element(ENABLE_NODE, &clear_labelpool_perf_test_cmd);
 #endif /* BGP_LABELPOOL_ENABLE_TESTS */
 }
+
+DEFINE_MTYPE_STATIC(BGPD, LABEL_PER_NEXTHOP_CACHE,
+		    "BGP Label Per Nexthop entry");
+
+int bgp_label_per_nexthop_cache_compare(
+	const struct bgp_label_per_nexthop_cache *a,
+	const struct bgp_label_per_nexthop_cache *b)
+{
+	return prefix_cmp(&a->prefix, &b->prefix);
+}
+
+struct bgp_label_per_nexthop_cache *
+bgp_label_per_nexthop_new(struct bgp_label_per_nexthop_cache_head *tree,
+			  struct prefix *prefix)
+{
+	struct bgp_label_per_nexthop_cache *blnc;
+
+	blnc = XCALLOC(MTYPE_LABEL_PER_NEXTHOP_CACHE,
+		       sizeof(struct bgp_label_per_nexthop_cache));
+	blnc->tree = tree;
+	blnc->label = MPLS_INVALID_LABEL;
+	prefix_copy(&blnc->prefix, prefix);
+	LIST_INIT(&(blnc->paths));
+	bgp_label_per_nexthop_cache_add(tree, blnc);
+
+	return blnc;
+}
+
+static void
+_bgp_label_per_nexthop_free(struct bgp_label_per_nexthop_cache *blnc)
+{
+	struct nexthop *nhop;
+
+	/* free MPLS label */
+	if (blnc->label == MPLS_LABEL_NONE || blnc->label == MPLS_INVALID_LABEL)
+		return;
```

Comment:
We allocate blnc in bgp_label_per_nexthop_new() and the label is MPLS_INVALID_LABEL. Here we check for this label, does it mean we never free the memory in such a case?

Reply:
this function is not the free memory function.
this just gives the MPLS label back to the system.

Reply:
ok

Reply:
done.